### PR TITLE
Ensuring extensions/blocks do not inadvertently register twice

### LINKED ISSUE #11506: Block type "jetpack/related-posts" is already registered
Recent versions of Jetpack are filling my debug logs with thousands of the following error:

```
Block type "jetpack/related-posts" is already registered. Please see <a href="https://codex.wordpress.org/Debugging_in_WordPress">Debugging in WordPress</a> for more information. (This message was added in version 5.0.0.) in /var/www/mysite.com/wp-includes/functions.php on line 4667
```
Not sure if this is relevant, but we are not using Gutenberg blocks, nor do we use any default display options for Related posts on our site. 

Related Posts is enabled and we retrieve a list of related posts using Jetpack_RelatedPosts::init_raw and then render the results ourselves.

### LINKED ISSUE #11590: Simple payments block registers twice
When sync tries to sync callables, it hits simple payments block registration and registration fails yelding a PHP notification, filling the PHP error log. :-(

1. when sync was trying to sync this callable https://github.com/Automattic/jetpack/blob/813a646e55f93d68f7f21f77254d2640fe83786a/sync/class.jetpack-sync-defaults.php#L246
2. that hits `jetpack_register_gutenberg_extensions` action…
3. …which triggers simple payments block registration: https://github.com/Automattic/jetpack/blob/813a646e55f93d68f7f21f77254d2640fe83786a/modules/simple-payments/simple-payments.php#L43

```
[25-Feb-2019 15:40:55 UTC] PHP Notice:  WP_Block_Type_Registry::register was called <strong>incorrectly</strong>. Block type "jetpack/simple-payments" is already registered. Please see <a href="https://codex.wordpress.org/Debugging_in_WordPress">Debugging in WordPress</a> for more information. (This message was added in version 5.0.0.) in /var/www/html/wp-includes/functions.php on line 4231
[25-Feb-2019 15:40:55 UTC] PHP Stack trace:
[25-Feb-2019 15:40:55 UTC] PHP   1. shutdown_action_hook() /var/www/html/wp-includes/load.php:0
[25-Feb-2019 15:40:55 UTC] PHP   2. do_action() /var/www/html/wp-includes/load.php:684
[25-Feb-2019 15:40:55 UTC] PHP   3. WP_Hook->do_action() /var/www/html/wp-includes/plugin.php:453
[25-Feb-2019 15:40:55 UTC] PHP   4. WP_Hook->apply_filters() /var/www/html/wp-includes/class-wp-hook.php:310
[25-Feb-2019 15:40:55 UTC] PHP   5. Jetpack_Sync_Sender->do_sync() /var/www/html/wp-includes/class-wp-hook.php:286
[25-Feb-2019 15:40:55 UTC] PHP   6. Jetpack_Sync_Sender->do_sync_and_set_delays() /var/www/html/wp-content/plugins/jetpack/sync/class.jetpack-sync-sender.php:106
[25-Feb-2019 15:40:55 UTC] PHP   7. Jetpack_Sync_Sender->do_sync_for_queue() /var/www/html/wp-content/plugins/jetpack/sync/class.jetpack-sync-sender.php:124
[25-Feb-2019 15:40:55 UTC] PHP   8. do_action() /var/www/html/wp-content/plugins/jetpack/sync/class.jetpack-sync-sender.php:199
[25-Feb-2019 15:40:55 UTC] PHP   9. WP_Hook->do_action() /var/www/html/wp-includes/plugin.php:453
[25-Feb-2019 15:40:55 UTC] PHP  10. WP_Hook->apply_filters() /var/www/html/wp-includes/class-wp-hook.php:310
[25-Feb-2019 15:40:55 UTC] PHP  11. Jetpack_Sync_Module_Callables->maybe_sync_callables() /var/www/html/wp-includes/class-wp-hook.php:286
[25-Feb-2019 15:40:55 UTC] PHP  12. Jetpack_Sync_Module_Callables->get_all_callables() /var/www/html/wp-content/plugins/jetpack/sync/class.jetpack-sync-module-callables.php:226
[25-Feb-2019 15:40:55 UTC] PHP  13. array_map() /var/www/html/wp-content/plugins/jetpack/sync/class.jetpack-sync-module-callables.php:86
[25-Feb-2019 15:40:55 UTC] PHP  14. Jetpack_Sync_Module_Callables->get_callable() /var/www/html/wp-content/plugins/jetpack/sync/class.jetpack-sync-module-callables.php:86
[25-Feb-2019 15:40:55 UTC] PHP  15. Jetpack_Gutenberg::get_availability() /var/www/html/wp-content/plugins/jetpack/sync/class.jetpack-sync-module-callables.php:93
[25-Feb-2019 15:40:55 UTC] PHP  16. do_action() /var/www/html/wp-content/plugins/jetpack/class.jetpack-gutenberg.php:350
[25-Feb-2019 15:40:55 UTC] PHP  17. WP_Hook->do_action() /var/www/html/wp-includes/plugin.php:453
[25-Feb-2019 15:40:55 UTC] PHP  18. WP_Hook->apply_filters() /var/www/html/wp-includes/class-wp-hook.php:310
[25-Feb-2019 15:40:55 UTC] PHP  19. Jetpack_Simple_Payments->register_gutenberg_block() /var/www/html/wp-includes/class-wp-hook.php:286
[25-Feb-2019 15:40:55 UTC] PHP  20. jetpack_register_block() /var/www/html/wp-content/plugins/jetpack/modules/simple-payments/simple-payments.php:65
[25-Feb-2019 15:40:55 UTC] PHP  21. register_block_type() /var/www/html/wp-content/plugins/jetpack/class.jetpack-gutenberg.php:29
[25-Feb-2019 15:40:55 UTC] PHP  22. WP_Block_Type_Registry->register() /var/www/html/wp-includes/blocks.php:27
[25-Feb-2019 15:40:55 UTC] PHP  23. _doing_it_wrong() /var/www/html/wp-includes/class-wp-block-type-registry.php:78
[25-Feb-2019 15:40:55 UTC] PHP  24. trigger_error() /var/www/html/wp-includes/functions.php:4231
```

Convo: p1551109306318900-slack-jetpack-gutenberg

This could be solved by moving Simple payments block registration to independent PHP file in [`extensions/blocks`](https://github.com/Automattic/jetpack/tree/4306205514b98a4c3362c8e5a30f03876bee140c/extensions/blocks). See https://github.com/Automattic/jetpack/issues/11506#issuecomment-470500551

Related to similar issue with Related posts block: https://github.com/Automattic/jetpack/issues/11506

### PR DESCRIPTION
**Pull request by gititon**
This PR ensures extensions/blocks cannot be registered more than once.

Fixes #11590 
Fixes #11506

#### Changes proposed in this Pull Request:

Will not register an extension/block if it's already registered.

#### Testing instructions:

See https://github.com/Automattic/jetpack/issues/11590#issue-422334893
Also, an additional phpunit test checks this functionality

#### Proposed changelog entry for your changes:
No changelog entry needed

### CODE REVIEW

File: sync/class.jetpack-sync-sender.php [hunk 1]
```diff
@@ -195,7 +195,7 @@ private function fastcgi_finish_request() {
 	}
 
 	public function do_sync_for_queue( $queue ) {
-
+return;
```
Comment (jeherve) [inline sync/class.jetpack-sync-sender.php]: I think this was committed by mistake, maybe?

File: class.jetpack-gutenberg.php [hunk 2]
```diff
@@ -23,7 +23,9 @@ function jetpack_register_block( $slug, $args = array() ) {
 		_doing_it_wrong( 'jetpack_register_block', 'Prefix the block with jetpack/ ', '7.1.0' );
 		$slug = 'jetpack/' . $slug;
 	}
-	return register_block_type( $slug, $args );
+
+	// Checking whether block is registered to ensure it isn't registered twice.
+	return Jetpack_Gutenberg::is_registered( $slug ) ? false : register_block_type( $slug, $args );
```
Comment (jeherve) [inline class.jetpack-gutenberg.php]: Any specific reason to do it all on one line in this case? I don't mind either way, but I think we tend to usually be more verbose / explicit and split those things onto 2 lines. what do you think?

**Review by jeherve — CHANGES_REQUESTED** [review]
Awesome. This may also fix #11506.

I think you have a few changes in here that did not belong in this PR though. Could you take a look?

Thanks!

File: class.jetpack-gutenberg.php [hunk 3]
```diff
@@ -384,6 +390,19 @@ public static function get_availability() {
...
+	/**
+	 * Check if an extension/block is already registered
+	 *
+	 * @since 7.2
+	 *
+	 * @param string $extension Name of extension to check.
+	 *
+	 * @return bool
+	 */
+	public static function is_registered( $extension ) {
```
Comment (simison) [inline class.jetpack-gutenberg.php]: Should we use `$slug` here instead of `$extensions` for consistency?

https://github.com/Automattic/jetpack/blob/9d7f20797168a2c62770f34b2e5f3f22a813fd5d/class.jetpack-gutenberg.php#L21

File: sync/class.jetpack-sync-sender.php (hunk 1 above)
Comment (lezama) [inline sync/class.jetpack-sync-sender.php]:
```suggestion
```

File: class.jetpack-gutenberg.php [hunk 4]
```diff
@@ -23,6 +23,12 @@ function jetpack_register_block( $slug, $args = array() ) {
 		_doing_it_wrong( 'jetpack_register_block', 'Prefix the block with jetpack/ ', '7.1.0' );
 		$slug = 'jetpack/' . $slug;
 	}
+
+	// Checking whether block is registered to ensure it isn't registered twice.
+	if ( Jetpack_Gutenberg::is_registered( $slug ) ) {
+		return false;
+	}
```
Comment (ockham) [inline class.jetpack-gutenberg.php]: The thing is, this sort of check is kinda what `register_block_type()` does, and what results in the warning that we're trying to fix here (#11590) :sweat_smile: So in a way, we're just cutting in line before that check to silence that warning, when the underlying issue is that some consumer registers a block twice (which it shouldn't). So _ideally_, we'd find out what consumer that is and change it accordingly :grimacing:

Comment (jeherve) [reply class.jetpack-gutenberg.php]: > ideally, we'd find out what consumer that is and change it accordingly

There is a use-case here for example:
https://github.com/Automattic/jetpack/issues/11506#issuecomment-470330131

File: tests/php/bootstrap.php [hunk 5]
```diff
@@ -118,8 +118,12 @@ function in_running_uninstall_group() {
...
 // slowest running tests. See the configuration in phpunit.xml.dist
 // @todo Remove version check when 5.1 is the minimum WP version.
-if ( version_compare( $wp_version, '5.1', '>=' ) ) {
+if ( file_exists( $test_root . '/includes/listener-loader.php' ) ) {
+	// version 5.1 and higher could have this set.
 	require $test_root . '/includes/listener-loader.php';
 } else {
-	require $test_root . '/includes/speed-trap-listener.php';
+	if ( file_exists( require $test_root . '/includes/speed-trap-listener.php' ) ) {
+		echo'UPDATE YOUR WP CORE TESTS HELPERS!';
```
Comment (jeherve) [inline tests/php/bootstrap.php]:
```suggestion
		echo 'UPDATE YOUR WP CORE TESTS HELPERS!';
```

**Review by jeherve — CHANGES_REQUESTED** [review]
I'll let @ockham chime in with his opinion since he commented here:
https://github.com/Automattic/jetpack/pull/11618/files#r267412770

Until then, just a minor nitpicking / editor shouting.

File: tests/php/bootstrap.php (hunk 5 above)
Comment (kraftbj) [reply tests/php/bootstrap.php]: I'm less inclined to include this since it would only fire on WordPress 5.0 installs with our requirements. We'll eliminate this completely once 5.2 is out and 5.1 is our minimum. That said, fine since we will remove it soon enough.

File: class.jetpack-gutenberg.php (hunk 4 above)
Comment (simison) [reply class.jetpack-gutenberg.php]: Seems like both are valid concerns but given the time constraints, this seems like a valid way to go for now?

**Review by ockham — APPROVED** [review]
> I'll let @ockham chime in with his opinion since he commented here:
> https://github.com/Automattic/jetpack/pull/11618/files#r267412770

I had another look at @simison's great analysis at #11590 and commented there: https://github.com/Automattic/jetpack/issues/11590#issuecomment-475807604 

tl;dr: Let's go with this approach for now :smile:

File: class.jetpack-gutenberg.php (hunk 4 above)
Comment (ockham) [reply class.jetpack-gutenberg.php]: Yep: https://github.com/Automattic/jetpack/pull/11618#pullrequestreview-218003272

**Review by jeherve — APPROVED** [review]
